### PR TITLE
Try to synchronize operations within one update

### PR DESCRIPTION
**Pull request by paultsimura**
- Held on https://github.com/Expensify/App/pull/42772 which updates the version of Onyx and includes this fix.

### Details

This PR changes how we process the `Onyx.update` operation:
1. Enqueue all the operations per individual key;
2. Group the collection-related keys into a single `mergeCollection` call;
3. Process the rest of the individual keys with `merge` calls;

### Related Issues

https://github.com/Expensify/App/issues/37560

### Automated Tests

I've added tests that cover the `Onyx.update()` with multiple merge operations of the same keys to verify that all the changes are batched into a single `mergeCollection` operation.

1. Perform a complex `Onyx.update()` operation with multiple updates of the same item in `merge` and `mergeCollection` operations.
2. Verify the collection items are updated with all the changes combined.

_Operation example ref:_ https://github.com/Expensify/react-native-onyx/blob/eb17a5dd0eba6b7b27f1e92e0fc81863091ba6c8/tests/unit/onyxTest.js#L1046-L1103

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by paultsimura** [comment]
@tgolen I think I found a way to fix the issue of the out-of-order updates.

The main idea is the following: when one `update` call comes with multiple merge/set/mergeCollection operations, – do not process anything asynchronously. Instead, build stack all the operations by keys, and then persist everything in one `multiGet` + `multiSet` take.

This PR is super draft, it lacks the validations as well as the broadcasting of the updates, but I would appreciate your opinion on the idea as a whole.

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1597]: Could this be a `_.reduce` instead?

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1603]: Since a single operation can't be both a `SET` and a `MERGE` there might be a little cleaning up to this logic which could happen (maybe an early return in the `SET` case?)

**Review by tgolen — COMMENTED** [review]
I love the direction this is going in and I think your approach is a good one. I hope it holds up as you keep working on it!

File: lib/Onyx.js
Comment (paultsimura) [reply lib/Onyx.js:1603]: The logic behind this is the following: I'm putting the `merge` and `set` operations in an array of a certain object key in the order they appear in the update operation.
However, since the `set` operation fully overwrites the existing object, it makes sense to keep only the operations that come after the last `set` (if it is present).

Therefore, eventually, a single key's update queue will consist of an array of `merge` operations with (optionally) only the first one in the array being `set`: `[set?, merge, merge, ...]`.

Does it make sense?

File: lib/Onyx.js
Comment (paultsimura) [reply lib/Onyx.js:1597]: Yes, will update 👍

**Comment by paultsimura** [comment]
In general, it does work – I gave it a little test run.
However, there are a lot of different callbacks in the current implementation and I'd appreciate your guidance on a more general level @tgolen.

Calls like these:

https://github.com/Expensify/react-native-onyx/blob/7cbf836cc1c107e82da42f372b82c91e5ecd1640/lib/Onyx.js#L1511-L1516

https://github.com/Expensify/react-native-onyx/blob/7cbf836cc1c107e82da42f372b82c91e5ecd1640/lib/Onyx.js#L1153-L1154

And all the places where we call the `sendActionToDevTools`. With the new approach, if we treat `update` as a single transaction with the eventual `merge`, it will be much harder to differentiate `mergeCollection` from `merge`.

What broadcasting approach would you suggest using here? If I broadcast each key update separately – would the collection listener handle each update individually?
Should I instead broadcast all collection-related keys as a single collection update, and non-collection keys as individual updates?

File: lib/Onyx.js
Comment (tgolen) [reply lib/Onyx.js:1603]: Ah, OK. That makes sense. Adding some code comments could help with that.

Does this also work in a case like this where a single key might have multiple operations that are separated by other keys?

1. key1 - merge
2. key2 - merge
3. key1 - merge

That should result in a single merge instruction for `key1`.

Comment (paultsimura) [reply lib/Onyx.js:1603]: > Does this also work in a case like this where a single key might have multiple operations that are separated by other keys?
> 
> 1. key1 - merge
> 2. key2 - merge
> 3. key1 - merge
> 
> That should result in a single merge instruction for `key1`.

Yes, I'll perform some tests and attach different scenarios in a while

**Comment by tgolen** [comment]
> If I broadcast each key update separately – would the collection listener handle each update individually?

I believe this would work yes, and the collection listener will handle each individual update. This would be the safest approach against causing regressions.

> Should I instead broadcast all collection-related keys as a single collection update, and non-collection keys as individual updates?

The collection updates get a little tricky because some collection mappings will be expecting the callback to be triggered with the entire collection, while others might be expecting the callback to be triggered with each object in the collection. This is dependent on the `waitForCollectionCallback` mapping option.

I think I prefer the first method because that will make the least amount of behavior changes for Onyx.

**Comment by paultsimura** [comment]
This looks like a promising picture to me when coming online, 46 operations were batched into one `update`:

image

Some of the keys had up to 7 `merge` operations stacked:

image

But eventually, the broadcast was called for each key only once, containing the final updated version of the object:
image

**Comment by paultsimura** [comment]
**_I have read the CLA Document and I hereby sign the CLA_**

**Comment by paultsimura** [comment]
@tgolen please make a sanity check when you have a minute, along with [this comment](https://github.com/Expensify/react-native-onyx/pull/490#issuecomment-1977262222).

The only thing that bothers me is that one failing test. I don't fully understand the nature of those subscriptions, and based on the comment in code, that first call is expected to be with `null, undefined` params 🤔 
https://github.com/Expensify/react-native-onyx/blob/7cbf836cc1c107e82da42f372b82c91e5ecd1640/lib/Onyx.js#L915-L918

**Comment by tgolen** [comment]
Wow, that is a great improvement!

I don't think the test is too much to worry about. You can update the tests to change the `nth` to be `0` instead of `1` now. It makes sense in the context of the change you're making. 

Before your PR:
- [This](https://github.com/Expensify/react-native-onyx/blob/7cbf836cc1c107e82da42f372b82c91e5ecd1640/tests/unit/onyxTest.js#L919-L923) `Onyx.update()` would have triggered each subscriber twice (first with `null`, second with the actual value)

After your PR:
- The updates are squished into a single callback with the final value 🎉 

I think the important part of `null` that needs to be kept in mind (and you might want to add this to the unit tests with these changes) is that if I subscribe to a key that doesn't have any data stored on it, then the subscriber should be triggered once with `null` for a value. This allows the component to render and then the `null` value is substituted for `undefined` in `withOnyx` [here](https://github.com/Expensify/react-native-onyx/blob/7cbf836cc1c107e82da42f372b82c91e5ecd1640/lib/withOnyx.js#L322). That way, when the component gets the `undefined` prop, the `defaultProp` will be picked up and used.

Does that all make sense?

**Comment by paultsimura** [comment]
> Does that all make sense?

Yes, updated the test as you recommended – to verify that the first call for the non-existing keys is made with `undefined`.

**Comment by paultsimura** [comment]
@tgolen how terrible is the idea to make the following change?

```js
case METHOD.MERGE_COLLECTION: 
    _.each(value, (_value, _key) => promises.push(() => merge(_key, _value)));
    break; 
```
https://github.com/Expensify/react-native-onyx/blob/69b2d5973b4aef19ce783440df55dedb30d4efd2/lib/Onyx.js#L1560-L1562

From what I understand, it will trigger the re-render of components listening to a collection N times?

**Comment by tgolen** [comment]
We use `mergecollection` quite a bit so I would definitely be worried about performance issues with that.

**Comment by paultsimura** [comment]
@tgolen I _think_ I've made a working solution: it updates the data in the correct order, **but** it's a bit slower than the current version.

Would it be possible to request a look from the side at the potential bottlenecks and improvements, especially the cache-related operations?
This PR is risky because it changes one of the mostly used Onyx functions, so an extra pair of eyes would really be helpful here.

**Comment by tgolen** [comment]
OK, great. Let me ping some of our experts on Slack to see if they can have a look at this PR. You can also see my comment [here](https://github.com/Expensify/react-native-onyx/pull/485#pullrequestreview-1922833969) about how it can be reliably tested to help protect against major regressions.

File: lib/Onyx.js
Comment (kidroca) [inline lib/Onyx.js:1646]: A comment explaining why we delete here would be handy. It seems it's to prevent duplicate notification of subscribers, because `keysChanged` probably triggers individual key subscribers

**Review by kidroca — COMMENTED** [review]
This Pull Request represents a positive step forward in enhancing performance by batching updates into a single persist-to-storage operation.

**Considerations for follow-up work:**
- **Refactoring for Code Deduplication:** There's an opportunity to further refine our codebase by reducing duplicate logic. Specifically, methods like `set` could be simplified to merely forward calls to `Onyx.update` with the appropriate parameters. Similarly, `Onyx.merge` might be refactored to leverage `Onyx.update` for batched updates. These changes would streamline our code and potentially simplify future maintenance.

- **Optimizing Merge Logic:** The current implementation's use of `multiGet` to read current values seems to diverge from our optimized merge strategy that emphasizes delta merges. Previously, we focused on reading full values only as needed to facilitate updates via `keyChanged`, while ensuring that only the delta of the merge was persisted to storage. The observed logic in this PR, which appears to send back the full JSON rather than just the delta, might warrant a reevaluation to ensure it aligns with our efficiency goals.


The overall changes introduced by this PR are commendable and contribute to ongoing efforts to improve performance.

File: lib/Onyx.js
Comment (paultsimura) [reply lib/Onyx.js:1646]: Yes, exactly. So the main idea is that we collect all the keys that belong to collections (even if they were updated individually), and process them as one collection while processing the non-collection keys as individuals.
The `keysChanged` notifies the subscribers subscribed to the whole collection as well as those subscribed to individual keys. So we remove the individual collection keys to avoid double notifications.

File: lib/Onyx.js
Comment (paultsimura) [reply lib/Onyx.js:1597]: After a closer look, don't we eventually read all the values from cache/storage anyway before updating the cached values?

https://github.com/Expensify/react-native-onyx/blob/3f77c6bd2030979da3d581ea01fc9fd5e6d646e5/lib/Onyx.js#L1525-L1530

Comment (paultsimura) [reply lib/Onyx.js:1597]: Since we need to wait for all the Storage write operations to complete, and then perform the `cache.get()` for each key and merge in-memory anyway to update the cache, is there any benefit from the `JSON_PATCH` in the current implementation?

I see there already was a similar discussion [here](https://github.com/Expensify/react-native-onyx/pull/238#discussion_r1123966066), looks like it was not completely resolved 🤔

**Comment by paultsimura** [comment]
@kidroca could you please help me understand whether the `mergeQueue` was introduced to chain the merge operations within one `update` call, or if there's more to it? How broad is the "single tick" definition?

https://github.com/Expensify/react-native-onyx/blob/3f77c6bd2030979da3d581ea01fc9fd5e6d646e5/lib/Onyx.js#L23-L25

https://github.com/Expensify/react-native-onyx/blob/3f77c6bd2030979da3d581ea01fc9fd5e6d646e5/lib/Onyx.js#L1285-L1291

**Comment by paultsimura** [comment]
I just discovered a behavior that looks like a blocker to me. Explained in detail here: https://github.com/Expensify/react-native-onyx/issues/301#issuecomment-1987950312

File: lib/Onyx.js
Comment (kidroca) [reply lib/Onyx.js]: Yes, we do `get` the latest value and perform a full merge with it to update cache, but we only try to send the delta for the write to storage.
The only benefit seems is we're sparing sending the full data for the write operation (and in fact this is only on native on web we send the full data for the merge)

**Comment by kidroca** [comment]
> @kidroca could you please help me understand whether the `mergeQueue` was introduced to chain the merge operations within one `update` call, or if there's more to it? How broad is the "single tick" definition?

The merge queue was created way before the `update` interface existed, AFAIK it serves use cases where `Onyx.merge` is called multiple times from actions in the same tick, to be fair I don't know exactly when it succeeds in batching more than one item to merge, I guess it would be during init or updates

File: lib/Onyx.js
Comment (kidroca) [inline lib/Onyx.js:1637]: I'm interested in the decision-making process for using `mergeCollection` versus individual merges based on the count of items in the update.

A few questions arise from this approach:

1. **Rationale Behind Individual Merges for Single Items**: Could you provide more details on the decision to favor individual updates over `mergeCollection` when there's only one item in the update? Understanding the rationale could clarify whether it's a performance consideration, a matter of update specificity, or another reason.

2. **Notification to Collection Subscribers**: How does this approach impact the notification of collection subscribers, especially in scenarios where a key is added or removed? The decision to handle single and multiple item updates differently could affect how subscribers are notified, potentially leading to inconsistencies.

3. **Consistency in Handling Updates**: If bypassing `mergeCollection` for single-item updates achieves the intended functionality and subscriber notification, could a similar logic not be applied to multiple item updates? Understanding the distinction between handling one versus multiple items could help in assessing potential improvements or optimizations in the update mechanism.

Your insights on these points would enhance my understanding and ensure that the logic aligns with the intended behavior for collection updates.

File: lib/Onyx.js
Comment (kidroca) [inline lib/Onyx.js:1677]: It seems possible for `null` values to appear in positions other than index 0, which implies that `merge` operations are allowed to pass `null` as a merge value.

Instead of attempting to identify a `null` value here, wouldn't it be more effective to address it during the `enqueueMergeOperation`? If the value provided for a key is `null`, we could directly modify the update queue for this key to `updateQueue[key] = [null];`, effectively treating the update for the key as a `set` operation since the initial value would be overridden by `null`.

This approach would enable us to eliminate irrelevant merges that precede the `null` value, and also simplify the check here having to iterate all operations looking for `null`

File: tests/unit/onyxTest.js
Comment (kidroca) [inline tests/unit/onyxTest.js:1037]: Great work on adding more, refining existing tests and relocating them here

File: lib/Onyx.js
Comment (kidroca) [reply lib/Onyx.js:1677]: I speculate that the decision to forego `multiSet` and `multiMerge` methods was made to capitalize on the existing logic for notifying subscribers and updating cache. 

It's still possible to try and leverage the `multi` methods in a potential follow-up

**Review by kidroca — COMMENTED** [review]
Overall, this PR is in great shape. However, I have raised a concern regarding the merge collection process, detailed in a separate note. Additionally, I've provided some comments on the new changes introduced in this PR since my last review.

File: lib/Onyx.js
Comment (kidroca) [reply lib/Onyx.js:1640]: Thank you for your comprehensive explanations, @paultsimura.

I see the logic behind the current method and its potential for performance improvement. However, I'd like to offer a suggestion - not as a directive but as a point for consideration - that we might not need to differentiate between single-item and multiple-item updates right now. This thought is guided by a preference for simplicity, aiming to ease the review from others.

Perhaps the special handling for "1 collection key" scenarios could be tabled as an area for future exploration. This way, we can focus on the core functionality at present, leaving room for optimization down the line. Of course, this is just my perspective, and I'm open to further discussion or to hear other team members' thoughts on this matter.

**Comment by tgolen** [comment]
> @tgolen this looks like a pretty unpredictable hole to me since the mergeCollection basically makes the object-dependencies in hooks useless, and we use such dependencies all across the App.

I agree, that is a bad hole. Did you consider changing `mergeCollection()` so that subscribers are only triggered if the object changes (so it works like `merge()`)? I think that should be a reasonable change to make. I think it would also be fine to do that in a follow-up issue so that PR you are working on here doesn't have the scope increased.

I'm just getting caught up from being on vacation last week. Is there anything that I can help with for a next step here? How is the QA coming on the original PR and what is the status/ETA of getting this merged?

**Comment by tgolen** [comment]
I'm placing this on HOLD right now in order for QA to be done on the latest version of Onyx without more un-QAed PRs from being merged.

**Comment by blazejkustra** [comment]
@paultsimura I should be able to help with this! 😄 

[OnyxUtils.js migration](https://github.com/Expensify/react-native-onyx/pull/523) is close to a merge (@fabioh8010 is testing it in NewDot). 

Would it be possible to wait with this PR until the migration is merged? This could resolve some of your type problems + you won't have to care about declaration OnyxUtils.d.ts file.

**Comment by blazejkustra** [comment]
Update: [OnyxUtils migration](https://github.com/Expensify/react-native-onyx/pull/523#pullrequestreview-1996416637) is in it's final state, waiting to be merged 😄

**Comment by blazejkustra** [comment]
@paultsimura Update: [OnyxUtils](https://github.com/Expensify/react-native-onyx/pull/523#pullrequestreview-1996416637) was merged yesterday, we can remove HOLD from the title!

**Comment by blazejkustra** [comment]
@paultsimura Please have a look at the fix [here](https://github.com/Expensify/react-native-onyx/pull/533). Thanks for reporting!

File: lib/OnyxUtils.ts
Comment (blazejkustra) [reply lib/OnyxUtils.ts:113]: Instead of assertion here, let's validate that `collectionValues` is of type `string[]`, we could also just do this:
```ts
const collectionValues = Object.values(keys.COLLECTION ?? {}) as string[];
```

File: lib/OnyxUtils.ts
Comment (blazejkustra) [inline lib/OnyxUtils.ts:261]: Let's remove "types" from JSDoc

File: lib/OnyxUtils.ts
Comment (blazejkustra) [inline lib/OnyxUtils.ts:263]: Don't forget to add return types to every function, we could also enforce this with an eslint rule

File: lib/OnyxUtils.ts
Comment (blazejkustra) [inline lib/OnyxUtils.ts:1026]: +1 Thanks for noticing this regression!

File: lib/OnyxUtils.ts
Comment (blazejkustra) [inline lib/OnyxUtils.ts:1053]: Missing return type here as well

File: lib/Onyx.ts
Comment (blazejkustra) [inline lib/Onyx.ts:588]:
```suggestion
                if (OnyxUtils.isValidMergeCollection(key, value)) {
```

File: lib/OnyxUtils.ts
Comment (blazejkustra) [inline lib/OnyxUtils.ts:1053]: Also how about this declaration?
```suggestion
function isValidMergeCollection<TKey extends CollectionKeyBase>(collectionKey: TKey, collection: Record<string, unknown>): boolean {
```

We can use `Record<string, unknown>` here because this function verifies that `collection` is a valid argument for `enqueueMergeOperation` function

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by blazejkustra** [comment]
Sorry for not posting these comments earlier, I reviewed the PR TS-wise. In case of any questions let me know! 😄

**Comment by blazejkustra** [comment]
@paultsimura I'm not feeling too good today 🤒 @fabioh8010 could you help reviewing here?

**Review by fabioh8010 — APPROVED** [review]
LGTM from TS side!

**Comment by tgolen** [comment]
Sorry, I have not been following the Onyx releases very closely. Where does this PR stand and does anyone know if we are close to removing the HOLD on this and moving it forward?

**Comment by tgolen** [comment]
OK, thanks! I just merged that PR, so I'll take this one off hold.

File: lib/Onyx.ts
Comment (tgolen) [inline lib/Onyx.ts:657]: Instead of the name `acc`, please us a full variable name.

File: lib/Onyx.ts
Comment (tgolen) [inline lib/Onyx.ts:660]: What does "later" imply? Can you please update the comment to be more clear?

File: lib/Onyx.ts
Comment (tgolen) [inline lib/Onyx.ts:647]: Please add more context to this comment to answer _why_ this needs to happen.

File: lib/Onyx.ts
Comment (tgolen) [inline lib/Onyx.ts:602]: Please add some comments about what this queue is for, what it's format is, and how it should be used.

**Review by tgolen — CHANGES_REQUESTED** [review]
The code looks good. I have some requests for improved comments.

I believe that QA is currently doing a regression test against the latest version of Onyx. We should consider doing the same with this change before it is merged, since it is such a fundamental change. If you haven't done that before, let me know and I can help coordinate it.

**Review by srikarparsi — APPROVED** [review]
This lgtm but I'll leave it to @tgolen for the final review

**Comment by mallenexpensify** [comment]
@tgolen it looks like there are [a couple](https://github.com/Expensify/App/issues/34686) of [issues](https://github.com/Expensify/App/issues/40589) held on this PR (and this PR is held on `App onyx bump`?)

I'm unsure of how big of a priority this is but I noticed you're focused on another project all week. If this is a priority, does it make sense to assign someone else for review? If not... we can hang tight.

**Review by tgolen — DISMISSED** [review]

**Comment by tgolen** [comment]
@paultsimura Can you please fix the conflicts? Also, can you link to whatever issue or PR this is on HOLD for?

**Comment by tgolen** [comment]
@mallenexpensify I'm around this week, so I'll work on getting this merged. I don't think it needs reassigned at all.

**Comment by tgolen** [comment]
@paultsimura Can you fix the conflicts here and link to whatever issue or PR this is on HOLD for?

**Comment by tgolen** [comment]
Maybe @mountiny knows if there is a current PR for bumping Onyx? I don't think we track it anywhere, but you could maybe look through the titles of the open PRs to see if you can spot anything.

**Review by tgolen — APPROVED** [review]

**Comment by mountiny** [comment]
Yep we got [this one](https://github.com/Expensify/App/pull/42772) and in the final stages of getting it done hopefully, its been a big effort 😅  

We are also discussing in this canvas thread if you want to follow https://expensify.slack.com/archives/C05R2V5GM1S/p1718007552204639?thread_ts=1717193187.147939&cid=C05R2V5GM1S

**Comment by tgolen** [comment]
Great, thanks!

**Comment by tgolen** [comment]
It looks like the version bump PR was merged, so I think I can take the hold off of this one. We might want to still let it sit for a couple of days to ensure the version bump isn't reverted until any regressions.

@paultsimura if you want to fix the conflicts again in the meantime, we can look at merging this next week. Maybe Tuesday or Wednesday.

**Comment by mountiny** [comment]
Yes! Luckily, it seems like no regressions have been found really in connection to the bump, which is... sus, but I welcome it 😂 

@paultsimura I think you can now work on getting this PR ready for review and re-test again.

**Comment by tgolen** [comment]
Bump @paultsimura

**Review by tgolen — DISMISSED** [review]

**Comment by tgolen** [comment]
OK, that's probably wise. In order to do that, you need to make an App PR that points to this specific branch. Then I can create an ad-hoc build that we can give to QA for full regression testing.

**Comment by tgolen** [comment]
That's the correct way to do it, yeah. There should be a couple other examples in the `package.json` which do this. I think, what I forgot to mention is that you need to temporarily include the built source files on your branch. I believe you can do this with the `git add -f` command (`-f` to force the file to be added and bypass the `.ignore` file).

**Comment by tgolen** [comment]
From the issue, it sounds like this can be closed for now.

**Review by tgolen — APPROVED** [review]
Looks great, thank you! Here we go!

**Comment by aldo-expensify** [comment]
@paultsimura with these changes, does `mergecollection` onyx update have a performance advantage vs a bunch of `merge` updates if they belong to the same collection? or it should be roughly the same because the separate `merge` updates will be put together as a `mergecollection`?

**Comment by aldo-expensify** [comment]
Nice, thanks!

**Comment by aldo-expensify** [comment]
@paultsimura I found a bug in the version of react-native-onyx we are currently using and it seems like your PR here is fixing it. The bug: https://github.com/Expensify/App/issues/46135#issue-2428111404

When are we planning to release the changes made here?

**Pull request merged**
